google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb

```diff
@@ -0,0 +1,284 @@
+      #
+      #   # When ready, stop the listen operation and close the stream.
+      #   listener.stop
+      #
+      class DocumentListener
+        include MonitorMixin
+
+        ##
+        # @private
+        # Creates the watch stream and listener object.
+        def initialize doc_ref, &callback
+          @doc_ref = doc_ref
+          raise ArgumentError if @doc_ref.nil?
+
+          @callback = callback
+          raise ArgumentError if @callback.nil?
+
+          @inventory    = []
+          @resume_token = nil
+          @read_time    = nil
+
+          super() # to init MonitorMixin
+        end
+
+        ##
+        # @private
+        # Starts the client listening for changes. This is called when the
+        # listener object is created, which is why this method is not part of
+        # the public API.
+        def start
+          synchronize { start_listening! }
+          self
+        end
+
+        ##
+        # Stops the client listening for changes.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   listener = nyc_ref.listen do |snapshot|
+        #     puts "The population of #{snapshot[:name]} "
+        #     puts "is #{snapshot[:population]}."
+        #   end
+        #
+        #   # When ready, stop the listen operation and close the stream.
+        #   listener.stop
+        #
+        def stop
+          synchronize do
+            @stopped = true
+            @request_queue.push self if @request_queue
+          end
+        end
+
+        ##
+        # Whether the client has stopped listening for changes.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   listener = nyc_ref.listen do |snapshot|
+        #     puts "The population of #{snapshot[:name]} "
+        #     puts "is #{snapshot[:population]}."
+        #   end
+        #
+        #   # Checks if the listener is stopped.
+        #   listener.stopped? #=> false
+        #
+        #   # When ready, stop the listen operation and close the stream.
+        #   listener.stop
+        #
+        #   # Checks if the listener is stopped.
+        #   listener.stopped? #=> true
+        #
+        def stopped?
+          synchronize { @stopped }
+        end
+
+        private
+
+        def send_callback doc_snp
+          @callback.call doc_snp
+        end
+
+        # rubocop:disable all
+
+        def background_run enum
+          loop do
+
+            # Break loop, close thread if stopped
+            break if synchronize { @stopped }
+
+            begin
+              # Cannot syncronize the enumerator, causes deadlock
```

Comment:
Yes, it could. Originally it was shared between the DocumentListener and QueryListener, but it was unfactored because of the churn during development. It can be extracted in a future refactoring, but I would like to make sure it is stable first. Duplication is easier to deal with than a bad abstraction.